File: apps/Applications/ios/Local For Vocal/Local For Vocal/Sources/Services/KafkaEventService.swift
import Combine
import Foundation

// MARK: - Kafka Event Service (SSE Client)

/// Connects to the backend SSE endpoint to receive real-time Kafka events.
/// Usage:
///   KafkaEventService.shared.connect(token: authToken)
///   KafkaEventService.shared.on(.orderStatusChanged) { event in ... }
@MainActor
public class KafkaEventService: ObservableObject {
    public static let shared = KafkaEventService()

    // MARK: - Published State
    @Published public var isConnected = false
    @Published public var lastEvent: KafkaEvent? = nil

    // MARK: - Event Publishers
    /// Global event stream — publishes ALL events
    public let eventStream = PassthroughSubject<KafkaEvent, Never>()

    // Per-type event publishers for targeted subscriptions
    private var typePublishers: [String: PassthroughSubject<KafkaEvent, Never>] = [:]

    // MARK: - Private State
    private var task: URLSessionDataTask? = nil
    private var urlSession: URLSession?
    private var authToken: String?
    private var retryCount = 0
    private let maxRetries = 5
    private var retryTimer: Timer?
    private var cancellables = Set<AnyCancellable>()

    private init() {}

    // MARK: - Connect

    /// Connect to the SSE endpoint. Call after login.
    public func connect(token: String) {
        self.authToken = token
        self.retryCount = 0
        startConnection()
    }

    /// Disconnect from SSE. Call on logout or app background.
    public func disconnect() {
        retryTimer?.invalidate()
        retryTimer = nil
        task?.cancel()
        task = nil
        urlSession?.invalidateAndCancel()
        urlSession = nil
        isConnected = false
        AppLogger.info("📡 KafkaEventService disconnected")
    }

    // MARK: - Event Subscriptions

    /// Subscribe to a specific event type
    public func on(_ type: KafkaEventType, handler: @escaping (KafkaEvent) -> Void)
        -> AnyCancellable
    {
        let key = type.rawValue
        if typePublishers[key] == nil {
            typePublishers[key] = PassthroughSubject<KafkaEvent, Never>()
        }
        return typePublishers[key]!
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: handler)
    }

    /// Subscribe to multiple event types
    public func on(_ types: [KafkaEventType], handler: @escaping (KafkaEvent) -> Void)
        -> [AnyCancellable]
    {
        return types.map { on($0, handler: handler) }
    }

    // MARK: - Private: SSE Connection

    private func startConnection() {
        guard let token = authToken else {
            AppLogger.error("❌ KafkaEventService: No auth token")
            return
        }

        let sseURL = URL(string: "\(AppEnvironment.current.sseEndpoint)")!
        var request = URLRequest(url: sseURL)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        request.timeoutInterval = TimeInterval.infinity  // SSE connections are long-lived

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = TimeInterval.infinity
        config.timeoutIntervalForResource = TimeInterval.infinity

        let session = URLSession(
            configuration: config, delegate: SSEDelegate(service: self), delegateQueue: nil)
        self.urlSession = session

        let dataTask = session.dataTask(with: request)
        self.task = dataTask
        dataTask.resume()

        AppLogger.info("📡 KafkaEventService connecting to SSE...")
    }

    // Called by the delegate when data arrives
    fileprivate func handleSSEData(_ data: Data) {
        guard let text = String(data: data, encoding: .utf8) else { return }

        // SSE format: "data: {json}\n\n" or "event: type\ndata: {json}\n\n"
        let lines = text.components(separatedBy: "\n")
        var currentData: String?

        for line in lines {
            if line.hasPrefix("data: ") {
                currentData = String(line.dropFirst(6))
            } else if line.isEmpty, let jsonString = currentData {
                // End of SSE event — parse it
                processEvent(jsonString)
                currentData = nil
            }
        }
    }

    private func processEvent(_ jsonString: String) {
        guard let data = jsonString.data(using: .utf8) else { return }

        do {
            let event = try JSONDecoder().decode(KafkaEvent.self, from: data)

            Task { @MainActor in
                self.lastEvent = event
                self.isConnected = true
                self.retryCount = 0
            }

            // Publish to global stream
            eventStream.send(event)

            // Publish to type-specific stream
            if let publisher = typePublishers[event.eventType] {
                publisher.send(event)
            }

            AppLogger.debug("📡 SSE event: \(event.eventType)")
        } catch {
            AppLogger.error("📡 Failed to decode SSE event: \(error)")
        }
    }

    fileprivate func handleConnectionError(_ error: Error) {
        Task { @MainActor in
            self.isConnected = false
        }

        // Don't retry if explicitly cancelled
        if (error as NSError).code == NSURLErrorCancelled { return }

        retryCount += 1
        if retryCount <= maxRetries {
            let delay = min(Double(retryCount * retryCount) * 2, 60)  // Exponential backoff, max 60s
            AppLogger.warning(
                "📡 SSE reconnecting in \(delay)s (attempt \(retryCount)/\(maxRetries))")

            DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
                self?.startConnection()
            }
        } else {
            AppLogger.error("📡 SSE max retries reached. Giving up.")
        }
    }
}

// MARK: - SSE URLSession Delegate

private class SSEDelegate: NSObject, URLSessionDataDelegate {
    weak var service: KafkaEventService?

    init(service: KafkaEventService) {
        self.service = service
    }

    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
        Task { @MainActor in
            service?.handleSSEData(data)
        }
    }

    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?)
    {
        if let error = error {
            Task { @MainActor in
                service?.handleConnectionError(error)
            }
        }
    }
}

File: apps/Applications/android/app/src/main/java/com/localforvocalstartup/app/data/services/KafkaEventService.kt
package com.localforvocalstartup.app.data.services

import com.localforvocalstartup.app.data.model.KafkaEvent
import com.localforvocalstartup.app.utils.AppLogger
import com.localforvocalstartup.app.utils.Environment
import com.google.gson.Gson
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.concurrent.TimeUnit

/**
 * Pure OkHttp SSE implementation — avoids requiring the okhttp-sse artifact separately.
 * Mirrors iOS KafkaEventService connection lifecycle.
 */
object KafkaEventService {
    private val gson = Gson()
    private val _eventStream = MutableSharedFlow<KafkaEvent>(extraBufferCapacity = 100)
    val eventStream: SharedFlow<KafkaEvent> = _eventStream.asSharedFlow()

    var isConnected = false
        private set

    private var token: String? = null
    private var retryCount = 0
    private const val MAX_RETRIES = 5
    private val scope = CoroutineScope(Dispatchers.IO)
    private var isRunning = false

    fun connect(authToken: String) {
        this.token = authToken
        this.retryCount = 0
        if (!isRunning) {
            isRunning = true
            scope.launch { startConnection() }
        }
    }

    fun disconnect() {
        isRunning = false
        isConnected = false
        AppLogger.info("📡 KafkaEventService disconnected")
    }

    private suspend fun startConnection() {
        val currentToken = token ?: run {
            AppLogger.error("❌ KafkaEventService: No auth token")
            return
        }

        val sseUrl = try { Environment.current.sseEndpoint } catch (e: Exception) { return }

        val client = OkHttpClient.Builder()
            .readTimeout(0, TimeUnit.MILLISECONDS)
            .retryOnConnectionFailure(true)
            .build()

        val request = Request.Builder()
            .url(sseUrl)
            .header("Authorization", "Bearer $currentToken")
            .header("Accept", "text/event-stream")
            .build()

        try {
            val response = client.newCall(request).execute()
            if (!response.isSuccessful) {
                AppLogger.error("📡 SSE failed: ${response.code}")
                handleRetry()
                return
            }
            isConnected = true
            retryCount = 0
            AppLogger.info("📡 KafkaEventService connected to SSE")

            val reader = BufferedReader(InputStreamReader(response.body?.byteStream()))
            val dataBuffer = StringBuilder()

            while (isRunning) {
                val line = reader.readLine() ?: break
                when {
                    line.startsWith("data:") -> dataBuffer.append(line.removePrefix("data:").trim())
                    line.isEmpty() && dataBuffer.isNotEmpty() -> {
                        try {
                            val event = gson.fromJson(dataBuffer.toString(), KafkaEvent::class.java)
                            _eventStream.emit(event)
                            AppLogger.debug("📡 SSE event: ${event.eventType}")
                        } catch (e: Exception) {
                            AppLogger.error("📡 Failed to decode SSE event: ${e.message}")
                        }
                        dataBuffer.clear()
                    }
                }
            }
        } catch (e: Exception) {
            isConnected = false
            AppLogger.error("📡 SSE connection error: ${e.message}")
            handleRetry()
        }
    }

    private suspend fun handleRetry() {
        if (retryCount >= MAX_RETRIES) {
            AppLogger.warning("📡 SSE max retries reached. Giving up.")
            isRunning = false
            return
        }
        retryCount++
        val delayMs = minOf(retryCount * retryCount * 2000L, 60000L)
        AppLogger.warning("📡 SSE reconnecting in ${delayMs / 1000}s (attempt $retryCount/$MAX_RETRIES)")
        delay(delayMs)
        startConnection()
    }
}
